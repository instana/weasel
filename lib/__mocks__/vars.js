export default {
  reportingUrl: 'http://eum.example.com',
  xhrTransmissionTimeout: 5432,
  beaconBatchingTime: 0
};
