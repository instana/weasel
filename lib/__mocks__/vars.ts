export default {
  reportingUrl: 'http://eum.example.com',
  xhrTransmissionTimeout: 5432,
  beaconBatchingTime: 0,
  secretPropertyKey: '__secret__',
  sessionId: undefined,
  sessionStorageKey: 'session',
  defaultSessionInactivityTimeoutMillis: 100,
  defaultSessionTerminationTimeoutMillis: 200,
  maxAllowedSessionTimeoutMillis:  500
};
