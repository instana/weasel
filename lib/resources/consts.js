export const urlMaxLength = 255;

export const initiatorTypes = {
  'other': 0,

  'img': 1,
  // IMAGE element inside a SVG
  'image': 1,

  'link': 2,
  'script': 3,
  'css': 4,

  'xmlhttprequest': 5,
  'fetch': 5,
  'beacon': 5,

  'html': 6,
  'navigation': 6
};

export const cachingTypes = {
  unknown: 0,
  cached: 1,
  validated: 2,
  fullLoad: 3
};
