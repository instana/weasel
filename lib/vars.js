// @flow

import {generateUniqueId, now} from './util';
import {performance} from './performance';

type Meta = {
  [key: string]: string
}

const defaultVars: {
  // This is the global object name which the user of the EUM
  // product has no influence over. This should be chosen
  // wisely such that conflicts with other tools are highly
  // unlikely.
  nameOfLongGlobal: string,

  // The trace id of the page load. You should never need to
  // change this.
  pageLoadTraceId: string,

  // An optional trace ID to correlate the page load trace
  // to a backend trace.
  // Set via:
  // eum('traceId', '123');
  pageLoadBackendTraceId: ?string,

  // All timestamps which are part of beacons will be adjusted
  // using this timestamp before transmission to save some bytes.
  // So what we are doing is:
  // timestampToTransmit = actualTimestamp - referenceTimestamp
  referenceTimestamp: number,

  // High resolution reference point. Will be used to relate
  // high res timestamps to a value that client and server
  // understands.
  highResTimestampReference: number,

  // The time at which the EUM initializer was executed. Useful
  // to get a load time indicator for older devices.
  initializerExecutionTimestamp: number,

  // Changes the URL to which beacons will be send.
  // Change via
  // eum('reportingUrl', '//eum.example.com');
  reportingUrl: ?string,

  // Defines an application identification mechanism. This value
  // will always be transfered with every beacon to associate
  // requests with a monitored system.
  // Change via:
  // eum('apiKey', 'myKey');
  apiKey: ?string,

  // Defines user-configurable application payloads. These payloads
  // will be transfered with the page load beacon and should be a
  // Set meta data via:
  // eum('meta', 'key', 'value');
  meta: Meta,

  // An array of URLs for which no data should be collected. This
  // information is currently only respected for XHR data collection.
  // Set via:
  // eum('ignoreUrls', [/example.com/]);
  ignoreUrls: RegExp[],

  // Whether or not ping like requests should be ignored from AJAX
  // collection. This is separate from ignoreUrls to handle cases
  // where users start monitoring using weasel and generate tons
  // of beacons which look like
  // GET /ping
  // These types of beacons are often times not really helpful and
  // users typically only learn about these types of beacons after
  // they have monitored using weasel for some time. At that point,
  // this could have generated hundreds of thousands of beacons
  // that nobody is interested in.
  //
  // Set via:
  // eum('ignorePings', false);
  ignorePings: boolean,

  // Timeout for the transmission of XHR beacons.
  //
  // Set via:
  // eum('xhrTransmissionTimeout', 60000);
  xhrTransmissionTimeout: number,

  // Whether or not the page should be considered as loaded using the
  // page load event. False to indicate that page loads should be
  // manually managed.
  //
  // THIS IS HIGHLY DISCOURAGED!
  //
  // When using the mechanism, ensure that you manually trigger the
  // page load event.
  //
  // Usage:
  // Activate manual page load tracking:
  // eum('manualPageLoadEvent');
  //
  // Trigger the manual page load
  // eum('triggerPageLoad');
  manualPageLoadEvent: boolean,
  manualPageLoadTriggered: boolean,

  // Whether or not weasel should automatically clear the resource timing
  // buffer after collecting resource timing information. This is used to
  // ensure that the buffer always has sufficient capacity for new
  // resource timing entries.
  //
  // Change via:
  // eum('autoClearResourceTimings', true | false);
  autoClearResourceTimings: boolean,

  // Name of the page which the user is currently viewing. May be set as
  // part of the page load or later for SPA, XHR and error beacons.
  // Any string can be used as the page name. Try to keep it short though
  // for efficiency sake.
  //
  // Set via:
  // eum('page', 'myPageName');
  page: ?string,

  // Sets the percentage of beacons and XHR requests which should be
  // sampled.
  // The range is 0..1. When set to '0.01', only 1% of the spans will be
  // sampled.
  // The sampling decision is propagated via the 'sp' = 0 | 1 beacon
  // parameter and the X-INSTANA-L HTTP header which is added to XHR
  // requests.
  //
  // Set via:
  // eum('sampleRate', 1);
  sampleRate: number
} = {
  nameOfLongGlobal: 'EumObject',
  pageLoadTraceId: generateUniqueId(),
  pageLoadBackendTraceId: null,
  referenceTimestamp: now(),
  highResTimestampReference: performance && performance.now ? performance.now() : 0,
  initializerExecutionTimestamp: now(),
  reportingUrl: null,
  apiKey: null,
  meta: {},
  ignoreUrls: [],
  ignorePings: true,
  xhrTransmissionTimeout: 20000,
  manualPageLoadEvent: false,
  manualPageLoadTriggered: false,
  autoClearResourceTimings: true,
  page: undefined,
  sampleRate: 1
};

export default defaultVars;
