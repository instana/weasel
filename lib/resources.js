// @flow

import {performance, isResourceTimingAvailable} from './performance';
import type {BeaconWithResourceTiming} from './types';
import {info, debug} from './debug';
import {createTrie} from './trie';
import {win} from './browser';
import vars from './vars';

// See https://w3c.github.io/resource-timing/
// See https://www.w3.org/TR/hr-time/

const urlMaxLength = 255;

const initiatorTypes = {
  'other': 0,

  'img': 1,
  // IMAGE element inside a SVG
  'image': 1,

  'link': 2,
  'script': 3,
  'css': 4,

  'xmlhttprequest': 5,
  'fetch': 5,
  'beacon': 5,

  'html': 6,
  'navigation': 6
};

const cachingTypes = {
  unknown: 0,
  cached: 1,
  validated: 2,
  fullLoad: 3
};

export function addResourceTimings(beacon: BeaconWithResourceTiming, minStartTime: ?number) {
  if (isResourceTimingAvailable && win.JSON) {
    const entries = getEntriesTransferFormat(performance.getEntriesByType('resource'), minStartTime);
    beacon['res'] = win.JSON.stringify(entries);

    if (vars.autoClearResourceTimings && performance.clearResourceTimings) {
      if (DEBUG) {
        debug('Automatically clearing resource timing buffer.');
      }
      performance.clearResourceTimings();
    }
  } else if (DEBUG) {
    info('Resource timing not supported.');
  }
}


export function getEntriesTransferFormat(performanceEntries: Array<Object>, minStartTime: ?number) {
  const trie = createTrie();

  for (let i = 0, len = performanceEntries.length; i < len; i++) {
    const entry = performanceEntries[i];
    if (minStartTime != null &&
        (entry['startTime'] - vars.highResTimestampReference + vars.referenceTimestamp) < minStartTime) {
      continue;
    }

    let url = entry.name;
    const lowerCaseUrl = url.toLowerCase();
    if (lowerCaseUrl === 'about:blank' || lowerCaseUrl.indexOf('javascript:') === 0) {
      continue;
    }

    if (url.length > urlMaxLength) {
      url = url.substring(0, urlMaxLength);
    }

    // We provide more detailed XHR insights via our XHR instrumentation.
    // The XHR instrumentation is available once the initialization was executed
    // (which is completely synchronous).
    if (entry['initiatorType'] !== 'xmlhttprequest' || entry['startTime'] < vars.highResTimestampReference) {
      trie.addItem(url, serializeEntry(entry));
    }
  }

  return trie.toJs();
}


function serializeEntry(entry: Object) {
  const result = [
    Math.round(entry['startTime'] - vars.highResTimestampReference),
    Math.round(entry['duration']),
    initiatorTypes[entry['initiatorType']] || initiatorTypes['other']
  ];

  // When timing data is available, we can provide additional information about
  // caching and resource sizes.
  if (typeof entry['transferSize'] === 'number' &&
      typeof entry['encodedBodySize'] === 'number' &&
      // All this information may not be available due to cross-origin.
      entry['transferSize'] > 0) {
    if (entry['transferSize'] === 0) {
      result.push(cachingTypes.cached);
    } else if (entry['transferSize'] > 0 && (entry['encodedBodySize'] === 0 || entry['transferSize'] < entry['encodedBodySize'])) {
      result.push(cachingTypes.validated);
    } else {
      result.push(cachingTypes.fullLoad);
    }

    result.push(entry['encodedBodySize'] || '');
    result.push(entry['decodedBodySize'] || '');
    result.push(entry['transferSize'] || '');
  } else {
    result.push('');
    result.push('');
    result.push('');
    result.push('');
  }

  if (entry['responseStart'] != null &&
      // timing allow origin check may have failed
      entry['responseStart'] >= entry['fetchStart']) {
    result.push(calculateTiming(entry['redirectEnd'], entry['redirectStart']));
    result.push(calculateTiming(entry['domainLookupStart'], entry['fetchStart']));
    result.push(calculateTiming(entry['domainLookupEnd'], entry['domainLookupStart']));
    if (entry['secureConnectionStart'] != null && entry['secureConnectionStart'] > 0) {
      result.push(calculateTiming(entry['secureConnectionStart'], entry['connectStart']));
      result.push(calculateTiming(entry['connectEnd'], entry['secureConnectionStart']));
    } else {
      result.push(calculateTiming(entry['connectEnd'], entry['connectStart']));
      result.push('');
    }
    result.push(calculateTiming(entry['responseStart'], entry['requestStart']));
    result.push(calculateTiming(entry['responseEnd'], entry['responseStart']));
  }

  return result.join(',')
    // remove empty trailing timings
    .replace(/,+$/, '');
}

function calculateTiming(a, b) {
  if (a == null || b == null ||
      // the values being equal indicates for example that a network connection didn't need
      // to be established. Do not report a timing of '0' as this will skew the statistics.
      a === b) {
    return '';
  }
  const diff = Math.round(a - b);
  if (diff < 0) {
    return '';
  }
  return diff;
}
