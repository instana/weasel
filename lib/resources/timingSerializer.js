// @flow

import { cachingTypes, initiatorTypes } from './consts';
import vars from '../vars';

export function serializeEntryToArray(entry: Object) {
  const result = [
    Math.round(entry['startTime'] - vars.highResTimestampReference),
    Math.round(entry['duration']),
    initiatorTypes[entry['initiatorType']] || initiatorTypes['other']
  ];

  // When timing data is available, we can provide additional information about
  // caching and resource sizes.
  if (typeof entry['transferSize'] === 'number' &&
      typeof entry['encodedBodySize'] === 'number' &&
      // All this information may not be available due to the timing allow origin check.
      entry['encodedBodySize'] > 0) {
    if (entry['transferSize'] === 0) {
      result.push(cachingTypes.cached);
    } else if (entry['transferSize'] > 0 && (entry['encodedBodySize'] === 0 || entry['transferSize'] < entry['encodedBodySize'])) {
      result.push(cachingTypes.validated);
    } else {
      result.push(cachingTypes.fullLoad);
    }

    if (entry['encodedBodySize'] != null) {
      result.push(entry['encodedBodySize']);
    } else {
      result.push('');
    }
    if (entry['decodedBodySize'] != null) {
      result.push(entry['decodedBodySize']);
    } else {
      result.push('');
    }
    if (entry['transferSize'] != null) {
      result.push(entry['transferSize']);
    } else {
      result.push('');
    }
  } else {
    result.push('');
    result.push('');
    result.push('');
    result.push('');
  }

  if (entry['responseStart'] != null &&
      // timing allow origin check may have failed
      entry['responseStart'] >= entry['fetchStart']) {
    result.push(calculateTiming(entry['redirectEnd'], entry['redirectStart']));
    result.push(calculateTiming(entry['domainLookupStart'], entry['fetchStart']));
    result.push(calculateTiming(entry['domainLookupEnd'], entry['domainLookupStart']));
    if (entry['secureConnectionStart'] != null && entry['secureConnectionStart'] > 0) {
      result.push(calculateTiming(entry['secureConnectionStart'], entry['connectStart']));
      result.push(calculateTiming(entry['connectEnd'], entry['secureConnectionStart']));
    } else {
      result.push(calculateTiming(entry['connectEnd'], entry['connectStart']));
      result.push('');
    }
    result.push(calculateTiming(entry['responseStart'], entry['requestStart']));
    result.push(calculateTiming(entry['responseEnd'], entry['responseStart']));
  }

  let backendTraceId = '';
  try {
    const serverTimings = entry['serverTiming'];
    if (serverTimings instanceof Array) {
      for (let i = 0; i < serverTimings.length; i++) {
        const serverTiming = serverTimings[i];
        if (serverTiming['name'] === vars.serverTimingBackendTraceIdEntryName) {
          backendTraceId = serverTiming['description'];
        }
      }
    }
  } catch (e) {
    // Some browsers may not grant access to the field when the Timing-Allow-Origin
    // check fails. Better be safe than sorry here.
  }
  result.push(backendTraceId);

  return result;
}

export function serializeEntry(entry: PerformanceEntry) {
  return serializeEntryToArray(entry).join(',')
    // remove empty trailing timings
    .replace(/,+$/, '');
}

function calculateTiming(a: ?number, b: ?number) {
  if (a == null || b == null ||
      // the values being equal indicates for example that a network connection didn't need
      // to be established. Do not report a timing of '0' as this will skew the statistics.
      a === b) {
    return '';
  }
  const diff = Math.round(a - b);
  if (diff < 0) {
    return '';
  }
  return diff;
}
