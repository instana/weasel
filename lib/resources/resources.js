// @flow

import {performance, isResourceTimingAvailable} from '../performance';
import type {BeaconWithResourceTiming} from '../types';
import {serializeEntry} from './timingSerializer';
import {urlMaxLength} from './consts';
import {info, debug} from '../debug';
import {createTrie} from '../trie';
import {win} from '../browser';
import vars from '../vars';

// See https://w3c.github.io/resource-timing/
// See https://www.w3.org/TR/hr-time/

export function addResourceTimings(beacon: BeaconWithResourceTiming, minStartTime: ?number) {
  if (isResourceTimingAvailable && win.JSON) {
    const entries = getEntriesTransferFormat(performance.getEntriesByType('resource'), minStartTime);
    beacon['res'] = win.JSON.stringify(entries);

    if (vars.autoClearResourceTimings && performance.clearResourceTimings) {
      if (DEBUG) {
        debug('Automatically clearing resource timing buffer.');
      }
      performance.clearResourceTimings();
    }
  } else if (DEBUG) {
    info('Resource timing not supported.');
  }
}


function getEntriesTransferFormat(performanceEntries: Array<Object>, minStartTime: ?number) {
  const trie = createTrie();

  for (let i = 0, len = performanceEntries.length; i < len; i++) {
    const entry = performanceEntries[i];
    if (minStartTime != null &&
        (entry['startTime'] - vars.highResTimestampReference + vars.referenceTimestamp) < minStartTime) {
      continue;
    }

    let url = entry.name;
    const lowerCaseUrl = url.toLowerCase();
    if (lowerCaseUrl === 'about:blank' || lowerCaseUrl.indexOf('javascript:') === 0) {
      continue;
    }

    if (url.length > urlMaxLength) {
      url = url.substring(0, urlMaxLength);
    }

    // We provide more detailed XHR insights via our XHR instrumentation.
    // The XHR instrumentation is available once the initialization was executed
    // (which is completely synchronous).
    if (entry['initiatorType'] !== 'xmlhttprequest' || entry['startTime'] < vars.highResTimestampReference) {
      trie.addItem(url, serializeEntry(entry));
    }
  }

  return trie.toJs();
}
