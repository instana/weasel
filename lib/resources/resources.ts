import {performance, isResourceTimingAvailable} from '../performance';
import { isTransmitionRequest } from '../transmission/util';
import type {BeaconWithResourceTiming} from '../types';
import {serializeEntry} from './timingSerializer';
import {stripSecrets} from '../stripSecrets';
import {isUrlIgnored} from '../ignoreRules';
import {urlMaxLength} from './consts';
import {createTrie} from '../trie';
import {win} from '../browser';
import {info} from '../debug';
import vars from '../vars';

// See https://w3c.github.io/resource-timing/
// See https://www.w3.org/TR/hr-time/

export function addResourceTimings(beacon: Partial<BeaconWithResourceTiming>, minStartTime?: number) {
  if (!!isResourceTimingAvailable && win.JSON) {
    const entries = getEntriesTransferFormat(performance.getEntriesByType('resource'), minStartTime);
    beacon['res'] = win.JSON.stringify(entries);
  } else if (DEBUG) {
    info('Resource timing not supported.');
  }
}


function getEntriesTransferFormat(performanceEntries: PerformanceEntryList, minStartTime?: number) {
  const trie = createTrie();

  for (let i = 0, len = performanceEntries.length; i < len; i++) {
    const entry = performanceEntries[i] as PerformanceResourceTiming;
    if (minStartTime != null &&
        (entry['startTime'] - vars.highResTimestampReference + vars.referenceTimestamp) < minStartTime) {
      continue;
    } else if (entry['duration'] < 0) {
      // Some old browsers do not properly implement resource timing. They report negative durations.
      // Ignore instead of reporting these, as the data isn't usable.
      continue;
    }

    let url = entry.name;
    if (isUrlIgnored(url)) {
      if (DEBUG) {
        info('Will not include data about resource because resource URL is ignored via ignore rules.', entry);
      }
      continue;
    }

    const lowerCaseUrl = url.toLowerCase();
    const initiatorType = entry['initiatorType'];
    if (lowerCaseUrl === 'about:blank' || lowerCaseUrl.indexOf('javascript:') === 0 || // some iframe cases
      // Data transmission can be visible as a resource. Do not report it.
      isTransmitionRequest(url)) {
      continue;
    }

    if (url.length > urlMaxLength) {
      url = url.substring(0, urlMaxLength);
    }

    // We provide more detailed XHR insights via our XHR instrumentation.
    // The XHR instrumentation is available once the initialization was executed
    // (which is completely synchronous).
    if ((initiatorType !== 'xmlhttprequest' && initiatorType !== 'fetch') || entry['startTime'] < vars.highResTimestampReference) {
      trie.addItem(stripSecrets(url), serializeEntry(entry));
    }
  }

  return trie.toJs();
}
