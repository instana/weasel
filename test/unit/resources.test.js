import { initiatorTypes, cachingTypes } from '../../lib/resources/consts';
import { serializeEntry } from '../../lib/resources/timingSerializer';
import vars from '../../lib/vars';

const columnMappings = [
  'time',
  'duration',
  'initiator',
  'cachingType',
  'encodedBodySize',
  'decodedBodySize',
  'transferSize',
  'redirect',
  'appcache',
  'dns',
  'tcp',
  'ssl',
  'request',
  'response',
  'backendTraceId',
  'ttfb'
];

const columnValueResolvers = {
  2: initiatorTypes,
  3: cachingTypes
};

describe('resources/timingSerializer', () => {
  beforeEach(() => {
    vars.highResTimestampReference = 0;
  });

  describe('serializeEntry', () => {
    it('must identify retrieval from cache', () => {
      const entry = {
        name: 'https://www.example.com/images/image.png',
        entryType: 'resource',
        startTime: 434.9000000001979,
        duration: 4.399999997986015,
        initiatorType: 'img',
        nextHopProtocol: 'http/2+quic/43',
        workerStart: 0,
        redirectStart: 0,
        redirectEnd: 0,
        fetchStart: 434.9000000001979,
        domainLookupStart: 434.9000000001979,
        domainLookupEnd: 434.9000000001979,
        connectStart: 434.9000000001979,
        connectEnd: 434.9000000001979,
        secureConnectionStart: 0,
        requestStart: 437.39999999888823,
        responseStart: 437.7999999996973,
        responseEnd: 439.2999999981839,
        transferSize: 0,
        encodedBodySize: 16786,
        decodedBodySize: 16786,
        serverTiming: []
      };
      const result = toHumanReadableEntry(serializeEntry(entry));
      expect(result).toMatchSnapshot();
      expect(result['backendTraceId']).toBe("");
    });

    it('must identify full asset retrieval', () => {
      const entry = {
        name: 'https://www.example.com/images/image.png',
        entryType: 'resource',
        startTime: 1136.8000000002212,
        duration: 231.5999999991618,
        initiatorType: 'img',
        nextHopProtocol: 'http/2+quic/43',
        workerStart: 0,
        redirectStart: 0,
        redirectEnd: 0,
        fetchStart: 1136.8000000002212,
        domainLookupStart: 1136.8000000002212,
        domainLookupEnd: 1136.8000000002212,
        connectStart: 1136.8000000002212,
        connectEnd: 1136.8000000002212,
        secureConnectionStart: 0,
        requestStart: 1141.7999999976018,
        responseStart: 1329.0999999990163,
        responseEnd: 1368.399999999383,
        transferSize: 16835,
        encodedBodySize: 16786,
        decodedBodySize: 16786,
        serverTiming: [
          {
            name: 'intid',
            description: '3748eabc1876cdaa'
          }
        ]
      };
      const result = toHumanReadableEntry(serializeEntry(entry));
      expect(result).toMatchSnapshot();
      expect(result['backendTraceId']).not.toBeNull();
    });

    it('must identify cache validation', () => {
      const entry = {
        name: 'https://www.example.com/images/image.png',
        entryType: 'resource',
        startTime: 1136.8000000002212,
        duration: 231.5999999991618,
        initiatorType: 'img',
        nextHopProtocol: 'http/2+quic/43',
        workerStart: 0,
        redirectStart: 0,
        redirectEnd: 0,
        fetchStart: 1136.8000000002212,
        domainLookupStart: 1136.8000000002212,
        domainLookupEnd: 1136.8000000002212,
        connectStart: 1136.8000000002212,
        connectEnd: 1136.8000000002212,
        secureConnectionStart: 0,
        requestStart: 1141.7999999976018,
        responseStart: 1329.0999999990163,
        responseEnd: 1368.399999999383,
        transferSize: 4320,
        encodedBodySize: 16786,
        decodedBodySize: 16786,
        serverTiming: []
      };
      const result = toHumanReadableEntry(serializeEntry(entry));
      expect(result).toMatchSnapshot();
      expect(result['backendTraceId']).toBe("");
    });
  });
});

function toHumanReadableEntry(str) {
  return str.split(',').reduce((agg, val, i) => {
    if (columnValueResolvers[i] && val !== '') {
      agg[columnMappings[i]] = getKeyForValue(columnValueResolvers[i], val) || 'UNKNOWN_EUM_VALUE';
    } else {
      agg[columnMappings[i]] = val;
    }
    return agg;
  }, {});
}

function getKeyForValue(obj, value) {
  const keys = Object.keys(obj);
  for (let i = 0; i < keys.length; i++) {
    const key = keys[i];
    if (String(obj[key]) === String(value)) {
      return key;
    }
  }
  return null;
}
