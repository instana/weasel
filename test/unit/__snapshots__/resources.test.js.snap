exports[`resources/timingSerializer serializeEntry must identify cache validation 1`] = `
Object {
  "appcache": "",
  "cachingType": "validated",
  "decodedBodySize": "16786",
  "dns": "",
  "duration": "232",
  "encodedBodySize": "16786",
  "initiator": "img",
  "redirect": "",
  "request": "187",
  "response": "39",
  "ssl": "",
  "tcp": "",
  "time": "1137",
  "transferSize": "4320",
}
`;

exports[`resources/timingSerializer serializeEntry must identify full asset retrieval 1`] = `
Object {
  "appcache": "",
  "cachingType": "fullLoad",
  "decodedBodySize": "16786",
  "dns": "",
  "duration": "232",
  "encodedBodySize": "16786",
  "initiator": "img",
  "redirect": "",
  "request": "187",
  "response": "39",
  "ssl": "",
  "tcp": "",
  "time": "1137",
  "transferSize": "16835",
}
`;

exports[`resources/timingSerializer serializeEntry must identify retrieval from cache 1`] = `
Object {
  "appcache": "",
  "cachingType": "cached",
  "decodedBodySize": "16786",
  "dns": "",
  "duration": "4",
  "encodedBodySize": "16786",
  "initiator": "img",
  "redirect": "",
  "request": "0",
  "response": "1",
  "ssl": "",
  "tcp": "",
  "time": "435",
  "transferSize": "0",
}
`;
